tools/pyright_tool.py

```diff
@@ -0,0 +1,658 @@
+            return False
+
+        for ignore_index, ignore in enumerate(self.all_pyright_ignores[file]):
+            if line_no == ignore.line_no:
+                for substring_index, ignore_statement in enumerate(
+                    ignore.ignore_statements
+                ):
+                    # Supporting both text substrings and regex patterns
+                    if ignore_statement.substring in error_message or re.search(
+                        ignore_statement.substring, error_message
+                    ):
+                        # Marking this ignore to be used (so we can identify unused ignores)
+                        self.all_pyright_ignores[file][ignore_index].ignore_statements[
+                            substring_index
+                        ].already_used = True
+                        return True
+
+        return False
+
+    def should_ignore_file_specific_error(self, file: str, error: Error) -> bool:
+        """Check if line should be ignored based on file-specific ignores."""
+        if file not in self.file_specific_ignores:
+            return False
+
+        for ignore_object in self.file_specific_ignores[file]:
+            if ignore_object.rule:
+                if error["rule"] == ignore_object.rule:
+                    ignore_object.already_used = True
+                    return True
+            elif ignore_object.substring:
+                # Supporting both text substrings and regex patterns
+                if ignore_object.substring in error["message"] or re.search(
+                    ignore_object.substring, error["message"]
+                ):
+                    ignore_object.already_used = True
+                    return True
+
+        return False
+
+    def is_line_in_pyright_off_block(self, file: str, line_no: int) -> bool:
+        """Check if line should be ignored based on `# pyright: off` mark."""
+        if file not in self.pyright_off_ignores:
+            return False
+
+        for off_ignore in self.pyright_off_ignores[file]:
+            if off_ignore.start_line < line_no < off_ignore.end_line:
+                off_ignore.already_used = True
+                return True
+
+        return False
+
+    def find_pyright_off_from_file(self, file: str) -> PyrightOffIgnores:
+        """Get sections in file to be ignored based on `# pyright: off`."""
+        pyright_off_ignores: PyrightOffIgnores = []
+        with open(file, "r") as f:
+            pyright_off = False
+            start_line = 0
+            index = 0
+            for index, line in enumerate(f):
+                if self.OFF_PATTERN in line:
+                    if not pyright_off:
+                        start_line = index
+                        pyright_off = True
+                elif self.ON_PATTERN in line:
+                    if pyright_off:
+                        pyright_off_ignores.append(PyrightOffIgnore(start_line, index))
+                        pyright_off = False
```

Comment:
```
if self.OFF_PATTERN in line and not pyright_off:
    ...
elif self.ON_PATTERN in line and pyright_off:
    ...
```
Would be a bit shorter but feel free to leave it as it is if you find it more readable.

Reply:
Good idea, it also gets rid of one indentation level